### DISCUSSION TITLE
How can DMS include a `X-Original-To` header (to support BCC when using DMS as a cache/proxy?)

### DISCUSSION THREAD

**Discussion by michelcve** [docker-mailserver/docker-mailserver#4056] [Q&A] [answered]
I'm using a docker-mailserver as a sort of "cache/proxy" mailserver.

This is running fine on a VPS that I hired somewhere. I use it for sending email, and for receiving it (and temporarily holding is).

It is configured with a catchall account, and appropriate alias so that all email for a single domain end up in 1 account.

My other mailserver (which runs at home) uses fetchmail to retrieve those emails, and process them. I have multiple users there, with aliases, (and also an alias set up with the same name as the user).

This setup works great for emails with the user in either TO or in CC.

However, for BCC this fails, and never gets delivered.

The problem is, that the "caching" mailserver does not add the postfix's "X-Original-To" (which makes sense for an normal mailserver, as you would want the BCC stripped there, being the whole point of a BCC of course).

I know there is a flag that can be used by postfix to still add the "X-Original-To" header when delivering mail to the final mailbox. Syntax for that would be something like this:

`dovecot unix - n n - - pipe flags=DROhu user=vmail:vmail argv=/usr/local/libexec/dovecot/deliver -f ${sender} -d ${recipient}`

Specifically the 'O' flag is what is needed.

Current configuration works a bit different, and doesnt use deliver, so it's not a matter of adding that O flag at the moment.

Anybody has any idea whether or not this is yet possible (or would be easy to "patch") for docker-mailserver?

My other alternative would be adding the same users and aliases to the cache server, but I would prefer to avoid that :-).

**Comment by glime-ai** [comment]
**Please note** that this response was created using generative AI, leveraging previously available information from the repository. This information may be helpful. However, it is our strong recommendation that you carefully consider the information before blindly applying it.

**This response is generated only once - you will not be able to interact with the glime-ai.**

---

To address the issue of BCC emails not being delivered due to the absence of the "X-Original-To" header, you can configure Postfix to add this header. This can be achieved by modifying the Postfix configuration to include the necessary flag.

Here is a step-by-step guide to achieve this:

1. **Edit the Postfix Configuration**:
   You need to modify the `master.cf` file to include the `X-Original-To` header. This can be done by adding a custom service in the `user-patches.sh` script.

2. **Create or Edit `user-patches.sh`**:
   Place the following script in your `docker-data/dms/config/` directory. If the file does not exist, create it.

   ```bash
   #!/bin/bash

   # Add the X-Original-To header
   postconf -e 'receive_override_options = no_address_mappings'
   postconf -e 'always_add_missing_headers = yes'
   postconf -e 'header_checks = regexp:/etc/postfix/header_checks'

   # Create the header_checks file
   echo '/^Received:/    IGNORE' > /etc/postfix/header_checks
   echo '/^X-Original-To:/    IGNORE' >> /etc/postfix/header_checks
   echo '/^To:/    IGNORE' >> /etc/postfix/header_checks
   echo '/^Cc:/    IGNORE' >> /etc/postfix/header_checks
   echo '/^Bcc:/    IGNORE' >> /etc/postfix/header_checks
   ```

3. **Make the Script Executable**:
   Ensure the `user-patches.sh` script is executable:

   ```bash
   chmod +x docker-data/dms/config/user-patches.sh
   ```

4. **Restart the Docker-Mailserver**:
   Restart the Docker-Mailserver container to apply the changes:

   ```bash
   docker-compose restart mailserver
   ```

This configuration will ensure that the `X-Original-To` header is added to the emails, allowing your fetchmail setup to correctly process BCC emails.

For more detailed information on using `user-patches.sh` and other advanced configurations, you can refer to the [official documentation](https://docker-mailserver.github.io/docker-mailserver/edge/config/advanced/override-defaults/user-patches).

This comment was generated by [Glime](https://glimelab.ai/?utm_source=0d5b86fb7eea4a54acac405212c29aba&utm_medium=github).

**Reply by michelcve** [reply]
Thanks for the tip AI, but unfortunately that doesn't work.

Now lets see if the humans here still outsmart the AI :-)

**Comment by polarathene** [comment, marked as answer]
I don't have time to review the past discussion on a related topic, but I did investigate this previously and provide insights + solution for enabling what you want (_along with caveats to account for_): https://github.com/orgs/docker-mailserver/discussions/3898

As the highlighted answer there notes, we need to sequentially deliver mail per recipient rather than delegating that for Dovecot to handle. That will allow the header you want to have added when used in combination with the additional flag.

Rather than calling the Dovecot deliver executable like your example shows, DMS uses LMTP to integrate with Dovecot. That should still be fine for your requirements AFAIK.

So try this `user-patches.sh` example to update `postfix-master.cf` during container start-up:

```bash
#!/bin/bash

# Required for the Postfix to send with original recipient headers for Sieve to use:
# Change from default limit of 50 to 1,
# Postfix will send mail with multiple recipients to Dovecot one at a time
# instead of delegating to Dovecot to manage all at once:
postconf lmtp_destination_recipient_limit=1

# Modify the `lmtp` service to use the `lmtp` transport command with extra `flags=DO` arg:
# This will enable both `Delivered-To:` and `X-Original-To` headers for Sieve to have access:
postconf -F 'lmtp/unix/command=lmtp flags=DO'
```

I've not reviewed anything else in the linked discussion, if that doesn't resolve it for you skim through the rest of the discussion and it might help you troubleshoot 👍

**Reply by michelcve** [reply]
That's great thanks! 

I gave it a try, and it works! I can see the `X-Original-To` header now being added to the messages.

Thanks a lot!
